### ISSUE TITLE
Shape Mismatch in MatMul During Forward Pass of ModernBertForSequenceClassification

### ISSUE THREAD

**Issue by whitebox2** [huggingface/candle#2961]
ModernBertForSequenceClassification model (hidden size = 768, sequence length = 128) to categorize text into one of classes. During the initial training epoch, however, the forward pass fails with a “shape mismatch in matmul” error.
Is there any way to solve this?

 #Error log
Tokenized shape: [4, 128]
Attention mask shape: [4, 128]
Input IDs shape: [4, 128]
Attention mask shape: [4, 128]
First sample token count: 128
Error in forward pass: shape mismatch in matmul, lhs: [4, 128], rhs: [768, 768]
Input shape: [4, 128], Attention mask shape: [4, 128]
Error: shape mismatch in matmul, lhs: [4, 128], rhs: [768, 768]

#Expected Behavior
    Input IDs should be a tensor of shape (batch_size, sequence_length) whose values are token indices (integers) and which the embedding layer then projects into the model’s hidden dimension (hidden_size = 768) before any matrix multiplication with weight matrices of shape (768, 768)
    The forward pass should succeed without dimension errors, yielding logits of shape (batch_size, num_classes).

#Code

```
use candle_core::{Device, Tensor, D, DType, Error};
use candle_nn::{ops, loss,  VarBuilder, optim::{Optimizer},var_map::VarMap};
use candle_transformers::models::modernbert::{ClassifierConfig, ClassifierPooling, ModernBertForSequenceClassification,Config
};
use hf_hub::{api::sync::Api, Repo, RepoType};
use tokenizers::{PaddingParams, Tokenizer};
use std::collections::HashMap;
use candle_optimisers::adam::{ParamsAdam, Adam};
use rand::{seq::SliceRandom, SeedableRng};
use rand::rngs::StdRng;
// Training settings
const LEARNING_RATE: f64 = 2e-5;
const EPOCHS: usize = 5;
const BATCH_SIZE: usize = 8;
const SEQ_LEN: usize = 128; // Sequence length
const SEED: u64 = 42;

// Data structure for text and label mapping
type LabeledDataset = HashMap<String, usize>;


fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Device selection (CPU or GPU)
    let device = candle_examples::device(true)?;
    println!("Using device: {:?}", device);
    
    // HuggingFace API configuration
    let revision = "main".to_string();
    let api = Api::new()?;
    let model_id = "answerdotai/ModernBERT-base".to_string();
    let repo = api.repo(Repo::with_revision(
        model_id,
        RepoType::Model,
        revision,
    ));

    // Load tokenizer and model configuration
    let tokenizer_filename = repo.get("tokenizer.json")?;
    let config_filename = repo.get("config.json")?;
    let weights_filename = repo.get("model.safetensors")?;
    
    // Load configuration file
    let config = std::fs::read_to_string(config_filename)?;
    let mut config: Config = serde_json::from_str(&config)?;
    
    // Output model configuration
    println!("Model config:");
    println!("  Hidden size: {}", config.hidden_size);
    println!("  Intermediate size: {}", config.intermediate_size);
    println!("  Max position embeddings: {}", config.max_position_embeddings);
    println!("  Num attention heads: {}", config.num_attention_heads);
    println!("  Num hidden layers: {}", config.num_hidden_layers);
    println!("  Vocab size: {}", config.vocab_size);

    
    // Check configuration compatibility
    if config.max_position_embeddings < SEQ_LEN {
        println!("Warning: SEQ_LEN ({}) is larger than max_position_embeddings ({}), adjusting SEQ_LEN",
                SEQ_LEN, config.max_position_embeddings);
    }
    
    // Initialize tokenizer
    let mut tokenizer = Tokenizer::from_file(tokenizer_filename).map_err(Error::msg)?;
    
    // Padding and truncation settings
    tokenizer
        .with_padding(Some(PaddingParams {
            strategy: tokenizers::PaddingStrategy::Fixed(SEQ_LEN),
            pad_id: config.pad_token_id,
            pad_token: "[PAD]".to_string(),
            pad_type_id: 0,
            pad_to_multiple_of: None,
            direction: tokenizers::PaddingDirection::Right,
        }))
        .with_truncation(Some(tokenizers::TruncationParams {
            max_length: SEQ_LEN,
            strategy: tokenizers::TruncationStrategy::LongestFirst,
            stride: 0,
            direction: tokenizers::TruncationDirection::Right,
        }))
        .map_err(Error::msg)?;

    // Configure label mappings
    let mut id2label = HashMap::new();
    let mut label2id = HashMap::new();

    let class_names = vec!["News", "Entertainment", "Sports", "Technology"];
    for (i, name) in class_names.iter().enumerate() {
        id2label.insert(i.to_string(), name.to_string());
        label2id.insert(name.to_string(), i.to_string());
    }
    
    // Add classifier configuration
    config.classifier_config = Some(ClassifierConfig {
        id2label: id2label.clone(),
        label2id: label2id.clone(),
        classifier_pooling: ClassifierPooling::CLS, // Use [CLS] token for pooling
    });

    // Create variable map for the model
    let mut varmap = VarMap::new();
    // Load model weights
    varmap.load(weights_filename)?;
    let vb = VarBuilder::from_varmap(&varmap, DType::F32, &device);
    
    // Display debug information before model instantiation
    println!("Input dimension compatibility check:");
    println!("  Expected embedding dimension: {}", config.hidden_size);
    println!("  Sequence length setting: {}", SEQ_LEN);
    
    // Model instantiation
    let model = match ModernBertForSequenceClassification::load(vb, &config) {
        Ok(m) => {
            println!("Model loaded successfully");
            m
        },
        Err(e) => {
            println!("Error loading model: {:?}", e);
            return Err(Box::new(e));
        }
    };

    // Optimizer parameter settings
    let params = ParamsAdam {
        lr: LEARNING_RATE as f64,
        ..Default::default()
    };
    

    let mut optimiser = Adam::new(varmap.all_vars(), params)?;

   
    let dataset = prepare_dataset();
    println!("Dataset size: {}", dataset.len());
    
    // Convert the HashMap to a vector for training
    let dataset_vec: Vec<(String, usize)> = dataset.into_iter().collect();
    println!("Dataset converted to vector with size: {}", dataset_vec.len());
    
    // Split into training and test data (80%:20%)
    let mut rng = StdRng::seed_from_u64(SEED);
    let mut dataset_indices: Vec<usize> = (0..dataset_vec.len()).collect();
    dataset_indices.shuffle(&mut rng);
    
    // Adjust the number of samples (to address potential issues)
    let train_size = std::cmp::min((dataset_vec.len() as f32 * 0.8) as usize, dataset_vec.len() - 1);
    let train_indices = &dataset_indices[0..train_size];
    let test_indices = &dataset_indices[train_size..];
    
    println!("Training set size: {}", train_indices.len());
    println!("Test set size: {}", test_indices.len());
    
    // Check the first data sample and output debug information
    if !dataset_vec.is_empty() {
        let sample_text = &dataset_vec[0].0;
        println!("Sample text: {}", sample_text);
        let encoded = tokenizer.encode(sample_text.as_str(), true).map_err(Error::msg)?;
        println!("Sample encoded length: {}", encoded.get_ids().len());
        println!("Sample tokens: {:?}", encoded.get_ids());
    }
    
    // Training by epoch
    for epoch in 0..EPOCHS {
        println!("Epoch: {}/{}", epoch + 1, EPOCHS);
        
        // Training phase
        let mut train_loss = 0.0;
        
        // Batch processing
        for batch_idx in (0..train_indices.len()).step_by(BATCH_SIZE) {
            let end = std::cmp::min(batch_idx + BATCH_SIZE, train_indices.len());
            let current_batch_size = end - batch_idx;
            
            // Prepare batch data
            let mut batch_texts = Vec::with_capacity(current_batch_size);
            let mut batch_labels = Vec::with_capacity(current_batch_size);
            
            for &idx in &train_indices[batch_idx..end] {
                batch_texts.push(dataset_vec[idx].0.as_str());
                batch_labels.push(dataset_vec[idx].1 as i64);
            }
            
            // Tokenization and get attention masks
            let input_ids = tokenize_batch(&tokenizer, batch_texts.clone(), &device)?;
            let attention_mask = get_attention_mask(&tokenizer, batch_texts.clone(), &device)?;
            
            // Output debug information (only for the first batch)
            if epoch == 0 && batch_idx == 0 {
                println!("Input IDs shape: {:?}", input_ids.shape());
                println!("Attention mask shape: {:?}", attention_mask.shape());
                println!("First sample token count: {}", 
                    tokenizer.encode(batch_texts[0], true).map_err(Error::msg)?.get_ids().len());
            }
            
            // Create label tensor
            let labels = Tensor::from_vec(batch_labels, current_batch_size, &device)?;
            
            // Model forward pass
            let logits = match model.forward(&input_ids, &attention_mask) {
                Ok(l) => l,
                Err(e) => {
                    // Display error details
                    println!("Error in forward pass: {:?}", e);
                    println!("Input shape: {:?}, Attention mask shape: {:?}", 
                             input_ids.shape(), attention_mask.shape());
                    return Err(Box::new(e));
                }
            };
            
            // Softmax and loss calculation
            let log_sm = ops::log_softmax(&logits, D::Minus1)?;
            let loss = loss::nll(&log_sm, &labels)?;
            let loss_val = loss.to_scalar::<f32>()?;
            train_loss += loss_val as f64 * current_batch_size as f64;
            
            // Backpropagation and optimization step
            optimiser.backward_step(&loss)?;
            
           
        }
        
      
    }
    
    println!("Training completed successfully!");
    
    // Save the model
    println!("Saving model...");
    let model_path = "model_weights.safetensors";
    varmap.save(model_path)?;
    println!("Model saved to {}", model_path);
   
    Ok(())
}


fn prepare_dataset() -> LabeledDataset {
  
    // 0: News, 1: Entertainment, 2: Sports, 3: Technology
    let mut dataset = HashMap::new();
    
    // News articles
    vec![
        "A new policy was announced in Tokyo. It aims to improve citizens' lives.",
        "An international conference is scheduled for next month. Many countries are expected to participate.",

    ].into_iter().for_each(|text| {
        dataset.insert(text.to_string(), 0);
    });
    
    // Entertainment
    vec![
        "A new movie was released over the weekend, achieving record attendance from day one.",
        "A popular artist's Japan tour was announced, and tickets sold out immediately.",

    ].into_iter().for_each(|text| {
        dataset.insert(text.to_string(), 1);
    });
    
    // Sports
    vec![
        "In a professional baseball game, a player hit a grand slam home run, leading the team to victory.",
 
    ].into_iter().for_each(|text| {
        dataset.insert(text.to_string(), 2);
    });
    
    // Technology
    vec![
        "A new smartphone was released, creating long lines at stores from day one.",

    ].into_iter().for_each(|text| {
        dataset.insert(text.to_string(), 3);
    });
    dataset
}
pub fn tokenize_batch(
    tokenizer: &Tokenizer,
    input: Vec<&str>,
    device: &Device,
) -> anyhow::Result<Tensor> {
    // Explicitly check tokenization settings
    println!("Tokenizing batch of {} texts", input.len());
    if !input.is_empty() {
        println!("Sample text length: {}", input[0].len());
    }
    
    let tokens = tokenizer.encode_batch(input, true).map_err(Error::msg)?;
    
    // Ensure input is not empty
    if tokens.is_empty() {
        return Err(anyhow::anyhow!("Tokenization returned empty result"));
    }
    
    // Check actual token length
    if !tokens.is_empty() {
        println!("Sample token length: {}", tokens[0].get_ids().len());
    }

    let token_ids = tokens
        .iter()
        .map(|tokens| {
            let mut ids = tokens.get_ids().to_vec();
            
            // Check if padding is needed
            if ids.len() < SEQ_LEN {
                // Pad with PAD token
                ids.resize(SEQ_LEN, 0);
            } else if ids.len() > SEQ_LEN {
                // Truncate if too long
                ids.truncate(SEQ_LEN);
            }
            
            Tensor::new(ids.as_slice(), device)
        })
        .collect::<candle_core::Result<Vec<_>>>()?;
    let result = Tensor::stack(&token_ids, 0)?;
    println!("Tokenized shape: {:?}", result.shape());
    Ok(result)
}

pub fn get_attention_mask(
    tokenizer: &Tokenizer,
    input: Vec<&str>,
    device: &Device,
) -> anyhow::Result<Tensor> {
    let tokens = tokenizer.encode_batch(input, true).map_err(Error::msg)?;

    let attention_mask = tokens
        .iter()
        .map(|tokens| {
            let mut masks = tokens.get_attention_mask().to_vec();
            
            // Adjust mask length
            if masks.len() < SEQ_LEN {
                // Fill shortfall with 0 (padding)
                masks.resize(SEQ_LEN, 0);
            } else if masks.len() > SEQ_LEN {
                // Truncate if too long
                masks.truncate(SEQ_LEN);
            }
            
            Tensor::new(masks.as_slice(), device)
        })
        .collect::<candle_core::Result<Vec<_>>>()?;
    
    let result = Tensor::stack(&attention_mask, 0)?;
    println!("Attention mask shape: {:?}", result.shape());
    Ok(result)
}
```

**Comment by whitebox2** [comment]
I might solve this issue.
The fix changes the indexing in ModernBertForSequenceClassification::forward():
-Before
```
impl ModernBertForSequenceClassification {
    pub fn forward(&self, xs: &Tensor, mask: &Tensor) -> Result<Tensor> {
        let output = self.model.forward(xs, mask)?;
        let last_hidden_state = match self.classifier_pooling {
            ClassifierPooling::CLS => output.i((.., .., 0))?, // ← this code
            // ...
        };
        // ...
    }
}
```
→
-After
```
impl ModernBertForSequenceClassification {
    pub fn forward(&self, xs: &Tensor, mask: &Tensor) -> Result<Tensor> {
        let output = self.model.forward(xs, mask)?;
        let last_hidden_state = match self.classifier_pooling {
            ClassifierPooling::CLS => output.i((.., 0, ..))?, // ← FIXED
            // ...
        };
        // ...
    }
}
```

#terminal
-Before
[ModernBertForSequenceClassification] last_hidden_state shape: Tensor[dims 4, 128; f32]
Error: shape mismatch in matmul, lhs: [4, 128], rhs: [768, 768]
→After
[ModernBertForSequenceClassification] last_hidden_state shape: [4, 768]

**Comment by whitebox2** [comment]
I found many problem in my code.So I will close.
